// @theia/preferences/lib/browser/views/components/preference-select-input.js
"use strict";
/********************************************************************************
 * Copyright (C) 2021 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreferenceSelectInputRenderer = void 0;
var preference_node_renderer_1 = require("./preference-node-renderer");
var inversify_1 = require("@theia/core/shared/inversify");
var PreferenceSelectInputRenderer = /** @class */ (function (_super) {
    __extends(PreferenceSelectInputRenderer, _super);
    function PreferenceSelectInputRenderer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PreferenceSelectInputRenderer.prototype.createInteractable = function (parent) {
        var e_1, _a;
        var options = this.preferenceNode.preference.data.enum;
        var interactable = document.createElement('select');
        this.interactable = interactable;
        interactable.classList.add('theia-select');
        interactable.onchange = this.handleUserInteraction.bind(this);
        try {
            for (var options_1 = __values(options), options_1_1 = options_1.next(); !options_1_1.done; options_1_1 = options_1.next()) {
                var value = options_1_1.value;
                var option = document.createElement('option');
                option.value = value;
                option.textContent = value;
                interactable.appendChild(option);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (options_1_1 && !options_1_1.done && (_a = options_1.return)) _a.call(options_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        interactable.value = this.getValue();
        parent.appendChild(interactable);
    };
    PreferenceSelectInputRenderer.prototype.getFallbackValue = function () {
        return this.preferenceNode.preference.data.enum[0];
    };
    PreferenceSelectInputRenderer.prototype.doHandleValueChange = function () {
        var currentValue = this.interactable.value || undefined;
        this.updateInspection();
        var newValue = this.getValue();
        this.updateModificationStatus(newValue);
        if (newValue !== currentValue && document.activeElement !== this.interactable) {
            this.interactable.value = newValue;
        }
    };
    PreferenceSelectInputRenderer.prototype.handleUserInteraction = function () {
        this.setPreferenceImmediately(this.interactable.value || undefined);
    };
    PreferenceSelectInputRenderer = __decorate([
        inversify_1.injectable()
    ], PreferenceSelectInputRenderer);
    return PreferenceSelectInputRenderer;
}(preference_node_renderer_1.PreferenceLeafNodeRenderer));
exports.PreferenceSelectInputRenderer = PreferenceSelectInputRenderer;
