// @theia/core/lib/browser/decorations-service.js
"use strict";
/********************************************************************************
 * Copyright (C) 2020 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecorationsServiceImpl = exports.DecorationsService = void 0;
var inversify_1 = require("inversify");
var common_1 = require("../common");
var ternary_search_tree_1 = require("../common/ternary-search-tree");
var uri_1 = require("../common/uri");
exports.DecorationsService = Symbol('DecorationsService');
var DecorationDataRequest = /** @class */ (function () {
    function DecorationDataRequest(source, thenable) {
        this.source = source;
        this.thenable = thenable;
    }
    return DecorationDataRequest;
}());
var DecorationProviderWrapper = /** @class */ (function () {
    function DecorationProviderWrapper(provider, onDidChangeDecorationsEmitter) {
        var _this = this;
        this.provider = provider;
        this.onDidChangeDecorationsEmitter = onDidChangeDecorationsEmitter;
        this.decorations = new Map();
        this.data = ternary_search_tree_1.TernarySearchTree.forUris(true);
        this.disposable = this.provider.onDidChange(function (uris) { return __awaiter(_this, void 0, void 0, function () {
            var uris_1, uris_1_1, uri, decoration, e_1_1;
            var e_1, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.decorations.clear();
                        if (!!uris) return [3 /*break*/, 1];
                        this.data.clear();
                        return [3 /*break*/, 8];
                    case 1:
                        _b.trys.push([1, 6, 7, 8]);
                        uris_1 = __values(uris), uris_1_1 = uris_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!uris_1_1.done) return [3 /*break*/, 5];
                        uri = uris_1_1.value;
                        this.fetchData(new uri_1.default(uri.toString()));
                        return [4 /*yield*/, provider.provideDecorations(uri, common_1.CancellationToken.None)];
                    case 3:
                        decoration = _b.sent();
                        if (decoration) {
                            this.decorations.set(uri.toString(), decoration);
                        }
                        _b.label = 4;
                    case 4:
                        uris_1_1 = uris_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (uris_1_1 && !uris_1_1.done && (_a = uris_1.return)) _a.call(uris_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 8:
                        this.onDidChangeDecorationsEmitter.fire(this.decorations);
                        return [2 /*return*/];
                }
            });
        }); });
    }
    DecorationProviderWrapper.prototype.dispose = function () {
        this.disposable.dispose();
        this.data.clear();
    };
    DecorationProviderWrapper.prototype.knowsAbout = function (uri) {
        return !!this.data.get(uri) || Boolean(this.data.findSuperstr(uri));
    };
    DecorationProviderWrapper.prototype.getOrRetrieve = function (uri, includeChildren, callback) {
        var item = this.data.get(uri);
        if (item === undefined) {
            // unknown -> trigger request
            item = this.fetchData(uri);
        }
        if (item && !(item instanceof DecorationDataRequest)) {
            // found something (which isn't pending anymore)
            callback(item, false);
        }
        if (includeChildren) {
            // (resolved) children
            var iter = this.data.findSuperstr(uri);
            if (iter) {
                var next = iter.next();
                while (!next.done) {
                    var value = next.value;
                    if (value && !(value instanceof DecorationDataRequest)) {
                        callback(value, true);
                    }
                    next = iter.next();
                }
            }
        }
    };
    DecorationProviderWrapper.prototype.fetchData = function (uri) {
        var _this = this;
        // check for pending request and cancel it
        var pendingRequest = this.data.get(new uri_1.default(uri.toString()));
        if (pendingRequest instanceof DecorationDataRequest) {
            pendingRequest.source.cancel();
            this.data.delete(uri);
        }
        var source = new common_1.CancellationTokenSource();
        var dataOrThenable = this.provider.provideDecorations(new uri_1.default(uri.toString()), source.token);
        if (!isThenable(dataOrThenable)) {
            // sync -> we have a result now
            return this.keepItem(uri, dataOrThenable);
        }
        else {
            // async -> we have a result soon
            var request_1 = new DecorationDataRequest(source, Promise.resolve(dataOrThenable).then(function (data) {
                if (_this.data.get(uri) === request_1) {
                    _this.keepItem(uri, data);
                }
            }).catch(function (err) {
                if (!(err instanceof Error && err.name === 'Canceled' && err.message === 'Canceled') && _this.data.get(uri) === request_1) {
                    _this.data.delete(uri);
                }
            }));
            this.data.set(uri, request_1);
            return undefined;
        }
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        function isThenable(obj) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            return obj && typeof obj.then === 'function';
        }
    };
    DecorationProviderWrapper.prototype.keepItem = function (uri, data) {
        var deco = data ? data : undefined;
        this.data.set(uri, deco);
        return deco;
    };
    return DecorationProviderWrapper;
}());
var DecorationsServiceImpl = /** @class */ (function () {
    function DecorationsServiceImpl() {
        this.data = [];
        this.onDidChangeDecorationsEmitter = new common_1.Emitter();
        this.onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
    }
    DecorationsServiceImpl.prototype.dispose = function () {
        this.onDidChangeDecorationsEmitter.dispose();
    };
    DecorationsServiceImpl.prototype.registerDecorationsProvider = function (provider) {
        var _this = this;
        var wrapper = new DecorationProviderWrapper(provider, this.onDidChangeDecorationsEmitter);
        this.data.push(wrapper);
        return common_1.Disposable.create(function () {
            // fire event that says 'yes' for any resource
            // known to this provider. then dispose and remove it.
            _this.data.splice(_this.data.indexOf(wrapper), 1);
            _this.onDidChangeDecorationsEmitter.fire(new Map());
            wrapper.dispose();
        });
    };
    DecorationsServiceImpl.prototype.getDecoration = function (uri, includeChildren) {
        var e_2, _a;
        var data = [];
        var containsChildren = false;
        try {
            for (var _b = __values(this.data), _c = _b.next(); !_c.done; _c = _b.next()) {
                var wrapper = _c.value;
                wrapper.getOrRetrieve(new uri_1.default(uri.toString()), includeChildren, function (deco, isChild) {
                    if (!isChild || deco.bubble) {
                        data.push(deco);
                        containsChildren = isChild || containsChildren;
                    }
                });
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return data;
    };
    DecorationsServiceImpl = __decorate([
        inversify_1.injectable()
    ], DecorationsServiceImpl);
    return DecorationsServiceImpl;
}());
exports.DecorationsServiceImpl = DecorationsServiceImpl;
